WIP: Add proxy functionality

pi0neerpat opened the pull request:
# Discussion

Is Library Lock modifier needed in addition to reentrancyGuard? I am not familiar with this, but according to the comments it says "@dev Prevents a contract from calling itself, directly or indirectly". 

[Example of library lock usage](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1822.md#token-logic-contract)

# Converting to proxy contract

In order to make this process easier, I created [sol-proxy](https://www.npmjs.com/package/sol-proxy). It's not fully finished, therefore the output needs some tweaking. Open to suggestions on how to improve this!

I already took care of these steps, but posting here just in case you want to run through it yourself.

### Generate Storage.sol
1. In `RToken.sol` remove  the `Storage.sol` import and the Storage inheritance. (you will add these back at the end)
2. `npm run proxify`
3. This will generate `Storage.sol` which will need some cleaning up. 

### Clean up Storage.sol
1. :warning: Most importantly, ensure the variables have not been re-ordered from a previously deployed version of the contract.
2. Remove duplicate pragma declarations. 
3. Add the necessary import statements.
4. All structs should be removed. If any of these structs are needed in Storage.sol, they must also be removed from their original contract and placed into `Structs.sol`.

hellwolf commented:
Can you elaborate:  Fix TypeError on npm run test (don't think thats related to my stuff) ?

pi0neerpat commented:
There is nothing wrong with the tests, this is something I was doing

hellwolf commented:
marking this related to #3

pi0neerpat commented:
All previously-written test are now passing.

## Changes made

### Variables and storage

Due to the way solidity handles variable declarations, duplicates will be ignored. Thus all variables are placed into the `Storage.sol` contract, which is inherited as the first base contract. This will make variables declared in `Storage.sol` as the "ultimate truth", and any further declaration instances are ignored.

*However*, rather than leave them in place, I decided to remove any obvious duplicate variable declarations, to reduce confusion around multiple declarations. This also will encourage future developers to only place variables in `Storage.sol` and to utilize the rules of variable ordering in proxy contracts.

__Change from Private to Public__: Since the variables are now instantiated in `Storage.sol`, they must be made public to be accessed by `RToken.sol`. One hiccup here is that making `Hat[] private hats` to "public" causes the following error

>TypeError: Internal or recursive type is not allowed for public state variables.

:warning: I have not found a real solution for this yet. The only hot-fix is to include the variable in both `Storage.sol` and `RToken.sol`

__Additional practices__ observed as suggested by OpenZeppelin ([article](https://docs.openzeppelin.com/sdk/2.5/writing-contracts)) and/or EIP 1822.


1. __Initializers rather than "constructor"__

  There is no need for a constructor, but rather a function which performs the similar effect as a constructor.

2. __Avoid initial values in field declarations__

  Since we are removing the use of a constructor, we also need to remove variables which are treated as "constructor-like", such as `myBoolean` in this example. If we do not do this, it will be difficult to change it's value in the future.

    ```solidity
    contract MyContract {
        bool myBoolean = true;

        constructor(){
        //...
        }
    }
    ```
  Instead, the above should be changed to this

    ```solidity
    contract MyContract {
        bool myBoolean;

        intilialize(){
          myBoolean = true;
        }
    }
    ```

  The following variables were impacted:

 `owner`: This can be changed with `transferOwnership()`, so it is not necessary to move this to the `initialize()`.

 `_guardCounter`: It is more likely we would implement a completely different version of ReentrancyGuard, rather than ever need to change this value. In this situation, `_guardCounter` would become deprecated / no longer used. Therefore, this was not moved to `initialize()`.

 `SELF_HAT_ID` and `PROPORTION_BASE`: The compiler does not reserve a storage slot for constants. Therefore, these were not moved to `initialize()`, nor placed in `Storage.sol`. If an upgraded contract is deployed with new constants, the proxy will utilize the new values.

 `name` `symbol` `decimals` `savingAssetConversionRate`: moved to `initialize()`, so they can be changed if needed during a future upgrade.

### Upgradeability

`RToken.sol` inherits `Proxiable.sol` to allow for upgrading

```solidity
contract RToken is Structs, Storage, IRToken, Ownable, Proxiable, LibraryLock, ReentrancyGuard {
```

`updateCode()` provides the pathway to perform an upgrade, restricted to only owner.

```solidity
/// @dev Update the rToken logic contract code
function updateCode(address newCode) external onlyOwner delegatedOnly {
  updateCodeAddress(newCode);
}
```

### Library Lock

Added to prevent potentially harmful calls made directly to the library/logic contract. For now, the only function which needs the `delegatedOnly` modifier is `updateCode()`.

```solidity
contract LibraryLock is Storage {
    // Ensures no one can manipulate the Logic Contract once it is deployed.
    // PARITY WALLET HACK PREVENTION

    modifier delegatedOnly() {
        require(initialized == true, "The library is locked. No direct 'call' is allowed");
        _;
    }
    function initialize() internal {
        initialized = true;
    }
}
```

### Other

__Don't include any `delegatecall` or `selfdestruct`__

  This is important to keep in mind for future upgrades. Never include this functionality, else an attacker can destroy the contracts.

__Compilation warnings__

```solidity
Warning: Unused local variable.
        (bool success, bytes memory _ ) = contractLogic.delegatecall(constructData); // solium-disable-line
                       ^------------^
```
This has been a warning on `Proxy.sol` for a while. Will need to defer to Gabriel if this can be fixed.


```solidity
Warning: Experimental features are turned on. Do not use experimental features on live deployments.
pragma experimental ABIEncoderV2;
^-------------------------------^
```

This was already present before starting my work.

## Todo

- [ ] Find out how to not need to declare `Hat[] private hats;` in `RToken.sol`
- [ ] Add a test to actually perform an upgrade, and repeat the test-suite again.

pi0neerpat commented:
## Upgrade test added & passing :+1:

Test #15 has been added for performing an upgrade. After the upgrade, it simply runs through test #2 again (rToken normal operations with zero hatter). If you want it to run through the entire test suite, simply place the upgrade snippet below in the `beforeEach()`.

```solidity
// Deploy the new rToken logic/library contract
const newRTokenLogic = await web3tx(RToken.new, "RToken.new")(
    {
      from: admin
  });
// Perform the upgrade
await web3tx(rToken.updateCode, "rToken.updateCode")(newRTokenLogic.address, {
  from: admin
});
```

### Other

`Ownable.sol` and `ReentrancyGuard.sol` each have their own "Private" variable declarations when using the imported contracts from OpenZeppelin. Since all the variables are listed as "Public" in `Storage.sol`, this is actually creating two separate variables, or at least it's unclear which is being used. Regardless, I copied the code from these two contracts to import them locally

```solidity
import {Ownable} from "openzeppelin-solidity/contracts/ownership/Ownable.sol";
import {ReentrancyGuard} from "openzeppelin-solidity/contracts/utils/ReentrancyGuard.sol";
// changed to
import {Ownable} from "./Ownable.sol";
import {ReentrancyGuard} from "./ReentrancyGuard.sol";
```

### Todo 

- [ ]  Fix TypeError on npm run test (don't think thats related to my stuff)

- [ ] Add LibraryLock modifier to function (pending)

- [ ] Does the initialize function need to be set to `delegatedOnly` ?

- [ ] Find out how to not need to declare Hat[] private hats; in RToken.sol

- [ ] Add a test to actually perform an upgrade, and repeat the test-suite again.

pi0neerpat commented on contracts/RToken.sol:
```diff
@@ -1,44 +1,39 @@
+import {Proxiable} from "./Proxiable.sol";
+import {LibraryLock} from "./LibraryLock.sol";
 import {SafeMath} from "openzeppelin-solidity/contracts/math/SafeMath.sol";
-import {Ownable} from "openzeppelin-solidity/contracts/ownership/Ownable.sol";
```
> Ownable.sol and ReentrancyGuard.sol each have their own "Private" variable declarations when using the imported contracts from OpenZeppelin. Since all the variables are listed as "Public" in Storage.sol, this is actually creating two separate variables, or at least it's unclear which is being used. Regardless, I copied the code from these two contracts to import them locally

Needed to strip out the variables from the OZ contract.

hellwolf replied:
I will leave this open, and I will consult a couple of ideas from others and look around examples... I will give myself a bit more time before giving up on this one.

pi0neerpat commented on contracts/RToken.sol:
```diff
@@ -1,44 +1,39 @@
-
+    uint32 public constant PROPORTION_BASE = 0xFFFFFFFF;
+    /// @dev Hat list
+    Hat[] private hats;
```
>    Not really a blocker. Just a little nit-pick on the clean-liness of the contracts.

No compilation warnings. I was unable to figure out how to declare this as public in `Storage.sol`. If you are able to figure it out, then we can certainly remove from `Rtoken.sol`

hellwolf replied:
yes, sorry, I should have read what you wrote!

hellwolf replied:
change private to internal :)

pi0neerpat replied:
Done! thanks

pi0neerpat commented on yarn.lock:
```diff
@@ -0,0 +1,4913 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Done, also added to .gitignore to prevent it in the future.

pi0neerpat commented on test/RToken.test.js:
```diff
@@ -1170,4 +1185,216 @@ contract("RToken contract", accounts => {
+            from: admin
+        });
+      // Perform the upgrade
+      await web3tx(rToken.updateCode, "rToken.updateCode")(newRTokenLogic.address, {
```
hmmmmm.... I can try to put the upgrade in the middle of a test. Would that satisfy the concern to demonstrate storage continuity between upgrades?

hellwolf replied:
yes please, that sounds goods.

pi0neerpat replied:
Done! Used test #2 again, but this time I put the upgrade in the middle. I had to modify the amounts of savings to account for the 2 extra blocks for 1) deploying the contract 2) performing the upgrade.

hellwolf commented on contracts/RToken.sol:
```diff
@@ -1,44 +1,39 @@
 pragma experimental ABIEncoderV2;
 
+import {Storage} from "./Storage.sol";
+import {Structs} from "./Storage.sol";
```
I'm a bit anal on this,  please make it `import {Structs} from "./Structs.sol"` <-- or I still insist changing it to RTokenStructs :)

pi0neerpat replied:
Done

pi0neerpat commented on contracts/Storage.sol:
```diff
@@ -0,0 +1,48 @@
+pragma solidity ^0.5.8;
```
done

pi0neerpat commented on contracts/Structs.sol:
```diff
@@ -0,0 +1,41 @@
+pragma solidity ^0.5.8;
+
+contract Structs {
```
done

hellwolf commented on flat.ignore.sol:
```diff
@@ -0,0 +1,1459 @@
+
```
hmm, what is this file??

pi0neerpat replied:
removed. It was a fragment from the sol-proxy, which flattens the contract before stripping out the variables

pi0neerpat commented on package.json:
```diff
@@ -1,27 +1,29 @@
-        "lint": "run-s lint:*",
-        "lint:js": "eslint . --max-warnings=0 --report-unused-disable-directives && echo '✔  Your .js files look good.'"
-    }
+  "name": "@decentral.ee/rtoken-contracts",
```
fixed. I had it off, since it was making huge changes on the RToken.test file.

pi0neerpat commented:
Adding a note to remind myself to look into the OZ initialization tool per the convo with quantstamp

hellwolf commented on contracts/Proxiable.sol:
```diff
@@ -0,0 +1,26 @@
+    }
+    function proxiableUUID() public pure returns (bytes32) {
+        return
+            0xc5f16f0fcc639fa48a6947836d9850f504798523bf8c9a3a87d5876cf622bcf7;
```
I found reference to this number on internet, we should use an unique number.

hellwolf commented on test/RToken.test.js:
```diff
@@ -5,6 +5,7 @@ const InterestRateModelMock = artifacts.require("InterestRateModelMock");
 const CErc20 = artifacts.require("CErc20");
```
please fix the lint errors in this file... my editor's correction if that helps:


```
diff --git a/test/RToken.test.js b/test/RToken.test.js
index 20d3b48..c1719db 100644
--- a/test/RToken.test.js
+++ b/test/RToken.test.js
@@ -56,16 +56,16 @@ contract("RToken contract", accounts => {
         // Deploy the rToken logic/library contract
         rTokenLogic = await web3tx(RToken.new, "RToken.new")(
             {
-              from: admin
-          });
+                from: admin
+            });
         // Get the init code for rToken
         const rTokenConstructCode = rTokenLogic.contract.methods.initialize(compoundAS.address).encodeABI();
 
         // Deploy the Proxy, using the init code for rToken
         const proxy = await web3tx(Proxy.new, "Proxy.new")(
-          rTokenConstructCode, rTokenLogic.address, {
-            from: admin
-          });
+            rTokenConstructCode, rTokenLogic.address, {
+                from: admin
+            });
         // Create the rToken object using the proxy address
         rToken = await RToken.at(proxy.address);
 
@@ -1187,216 +1187,216 @@ contract("RToken contract", accounts => {
     });
 
     it("#15 upgrade contract", async () => {
-      // Deploy the new rToken logic/library contract
-      const newRTokenLogic = await web3tx(RToken.new, "RToken.new")(
-          {
+        // Deploy the new rToken logic/library contract
+        const newRTokenLogic = await web3tx(RToken.new, "RToken.new")(
+            {
+                from: admin
+            });
+        // Perform the upgrade
+        await web3tx(rToken.updateCode, "rToken.updateCode")(newRTokenLogic.address, {
             from: admin
         });
-      // Perform the upgrade
-      await web3tx(rToken.updateCode, "rToken.updateCode")(newRTokenLogic.address, {
-        from: admin
-      });
-
-      // Below is just copy of test #2 rToken normal operations with zero hatter
-      await web3tx(token.approve, "token.approve 100 by customer1")(rToken.address, toWad(100), {
-          from: customer1
-      });
-      await web3tx(rToken.mint, "rToken.mint 100 to customer1", {
-          inLogs: [{
-              name: "Mint"
-          }, {
-              name: "Transfer",
-              args: {
-                  from: rToken.address,
-                  to: customer1,
-                  value: toWad(100)
-              }
-          }]
-      })(toWad(100), {
-          from: customer1
-      });
-      assert.equal(wad4human(await token.balanceOf.call(customer1)), "900.00000");
-      assert.equal(wad4human(await rToken.totalSupply.call()), "100.00000");
-      await expectAccount(customer1, {
-          tokenBalance: "100.00000",
-          cumulativeInterest: "0.00000",
-          receivedLoan: "100.00000",
-          receivedSavings: "100.00000",
-          interestPayable: "0.00000",
-      });
-
-      await expectRevert(rToken.transfer(customer2, toWad(100.1), { from: customer1 }), "Not enough balance to transfer");
-      await expectRevert(rToken.transferFrom(customer1, customer2, toWad(1), { from: admin }), "Not enough allowance for transfer");
-
-      await doBingeBorrowing();
-      await expectAccount(customer1, {
-          tokenBalance: "100.00000",
-          cumulativeInterest: "0.00000",
-          receivedLoan: "100.00000",
-          receivedSavings: "100.00100",
-          interestPayable: "0.00100",
-      });
-      assert.deepEqual(parseGlobalStats(await rToken.getGlobalStats.call()), {
-          totalSupply: "100.00000",
-          totalSavingsAmount: "100.00100"
-      });
-
-      await web3tx(rToken.redeem, "rToken.redeem 10 to customer1", {
-          inLogs: [{
-              name: "Redeem",
-              args: {
-                  redeemer: customer1,
-                  redeemTo: customer1,
-                  redeemAmount: toWad(10),
-              }
-          }, {
-              name: "Transfer",
-              args: {
-                  from: customer1,
-                  to: rToken.address,
-                  value: toWad(10)
-              }
-          }]
-      })(toWad(10), {
-          from: customer1
-      });
-      assert.equal(wad4human(await token.balanceOf.call(customer1)), "910.00000");
-      assert.equal(wad4human(await rToken.totalSupply.call()), "90.00101");
-      await expectAccount(customer1, {
-          tokenBalance: "90.00101",
-          cumulativeInterest: "0.00101",
-          receivedLoan: "90.00000",
-          receivedSavings: "90.00101",
-          interestPayable: "0.00000",
-      });
-
-      await web3tx(rToken.payInterest, "rToken.payInterest to customer1", {
-          inLogs: [{
-              name: "InterestPaid"
-          }, {
-              name: "Transfer",
-              args: {
-                  from: rToken.address,
-                  to: customer1
-                  // value: // who knows
-              }
-          }]
-      })(customer1, { from : admin });
-      assert.equal(wad4human(await token.balanceOf.call(customer1)), "910.00000");
-      assert.equal(wad4human(await rToken.totalSupply.call()), "90.00102");
-      await expectAccount(customer1, {
-          tokenBalance: "90.00102",
-          cumulativeInterest: "0.00102",
-          receivedLoan: "90.00000",
-          receivedSavings: "90.00102",
-          interestPayable: "0.00000",
-      });
-      await web3tx(rToken.payInterest, "rToken.payInterest to customer1 again", {
-          inLogs: [{
-              name: "InterestPaid"
-          }]
-      })(customer1, { from : admin });
-      await expectAccount(customer1, {
-          tokenBalance: "90.00103",
-          cumulativeInterest: "0.00103",
-          receivedLoan: "90.00000",
-          receivedSavings: "90.00103",
-          interestPayable: "0.00000",
-      });
-      assert.deepEqual(parseGlobalStats(await rToken.getGlobalStats.call()), {
-          totalSupply: "90.00103",
-          totalSavingsAmount: "90.00103"
-      });
-
-      await web3tx(rToken.redeemAndTransfer, "rToken.redeem 2 of customer1 to customer3", {
-          inLogs: [{
-              name: "Redeem",
-              args: {
-                  redeemer: customer1,
-                  redeemTo: customer3,
-                  redeemAmount: toWad(2),
-              }
-          }, {
-              name: "Transfer",
-              args: {
-                  from: customer1,
-                  to: rToken.address,
-                  value: toWad(2)
-              }
-          }]
-      })(customer3, toWad(2), {
-          from: customer1
-      });
-      assert.equal(wad4human(await token.balanceOf.call(customer3)), "2.00000");
-    })
+
+        // Below is just copy of test #2 rToken normal operations with zero hatter
+        await web3tx(token.approve, "token.approve 100 by customer1")(rToken.address, toWad(100), {
+            from: customer1
+        });
+        await web3tx(rToken.mint, "rToken.mint 100 to customer1", {
+            inLogs: [{
+                name: "Mint"
+            }, {
+                name: "Transfer",
+                args: {
+                    from: rToken.address,
+                    to: customer1,
+                    value: toWad(100)
+                }
+            }]
+        })(toWad(100), {
+            from: customer1
+        });
+        assert.equal(wad4human(await token.balanceOf.call(customer1)), "900.00000");
+        assert.equal(wad4human(await rToken.totalSupply.call()), "100.00000");
+        await expectAccount(customer1, {
+            tokenBalance: "100.00000",
+            cumulativeInterest: "0.00000",
+            receivedLoan: "100.00000",
+            receivedSavings: "100.00000",
+            interestPayable: "0.00000",
+        });
+
+        await expectRevert(rToken.transfer(customer2, toWad(100.1), { from: customer1 }), "Not enough balance to transfer");
+        await expectRevert(rToken.transferFrom(customer1, customer2, toWad(1), { from: admin }), "Not enough allowance for transfer");
+
+        await doBingeBorrowing();
+        await expectAccount(customer1, {
+            tokenBalance: "100.00000",
+            cumulativeInterest: "0.00000",
+            receivedLoan: "100.00000",
+            receivedSavings: "100.00100",
+            interestPayable: "0.00100",
+        });
+        assert.deepEqual(parseGlobalStats(await rToken.getGlobalStats.call()), {
+            totalSupply: "100.00000",
+            totalSavingsAmount: "100.00100"
+        });
+
+        await web3tx(rToken.redeem, "rToken.redeem 10 to customer1", {
+            inLogs: [{
+                name: "Redeem",
+                args: {
+                    redeemer: customer1,
+                    redeemTo: customer1,
+                    redeemAmount: toWad(10),
+                }
+            }, {
+                name: "Transfer",
+                args: {
+                    from: customer1,
+                    to: rToken.address,
+                    value: toWad(10)
+                }
+            }]
+        })(toWad(10), {
+            from: customer1
+        });
+        assert.equal(wad4human(await token.balanceOf.call(customer1)), "910.00000");
+        assert.equal(wad4human(await rToken.totalSupply.call()), "90.00101");
+        await expectAccount(customer1, {
+            tokenBalance: "90.00101",
+            cumulativeInterest: "0.00101",
+            receivedLoan: "90.00000",
+            receivedSavings: "90.00101",
+            interestPayable: "0.00000",
+        });
+
+        await web3tx(rToken.payInterest, "rToken.payInterest to customer1", {
+            inLogs: [{
+                name: "InterestPaid"
+            }, {
+                name: "Transfer",
+                args: {
+                    from: rToken.address,
+                    to: customer1
+                    // value: // who knows
+                }
+            }]
+        })(customer1, { from : admin });
+        assert.equal(wad4human(await token.balanceOf.call(customer1)), "910.00000");
+        assert.equal(wad4human(await rToken.totalSupply.call()), "90.00102");
+        await expectAccount(customer1, {
+            tokenBalance: "90.00102",
+            cumulativeInterest: "0.00102",
+            receivedLoan: "90.00000",
+            receivedSavings: "90.00102",
+            interestPayable: "0.00000",
+        });
+        await web3tx(rToken.payInterest, "rToken.payInterest to customer1 again", {
+            inLogs: [{
+                name: "InterestPaid"
+            }]
+        })(customer1, { from : admin });
+        await expectAccount(customer1, {
+            tokenBalance: "90.00103",
+            cumulativeInterest: "0.00103",
+            receivedLoan: "90.00000",
+            receivedSavings: "90.00103",
+            interestPayable: "0.00000",
+        });
+        assert.deepEqual(parseGlobalStats(await rToken.getGlobalStats.call()), {
+            totalSupply: "90.00103",
+            totalSavingsAmount: "90.00103"
+        });
+
+        await web3tx(rToken.redeemAndTransfer, "rToken.redeem 2 of customer1 to customer3", {
+            inLogs: [{
+                name: "Redeem",
+                args: {
+                    redeemer: customer1,
+                    redeemTo: customer3,
+                    redeemAmount: toWad(2),
+                }
+            }, {
+                name: "Transfer",
+                args: {
+                    from: customer1,
+                    to: rToken.address,
+                    value: toWad(2)
+                }
+            }]
+        })(customer3, toWad(2), {
+            from: customer1
+        });
+        assert.equal(wad4human(await token.balanceOf.call(customer3)), "2.00000");
+    });
 
     it("#16 proxy security", async () => {
 
-      // Call initialize first time
-      await web3tx(rTokenLogic.initialize, "rTokenLogic.initialize first time")(compoundAS.address, {
-          from: admin
-      })
+        // Call initialize first time
+        await web3tx(rTokenLogic.initialize, "rTokenLogic.initialize first time")(compoundAS.address, {
+            from: admin
+        });
 
-      // Test original logic contract
-      await expectRevert(web3tx(rTokenLogic.initialize, "rTokenLogic.initialize second time")(compoundAS.address, {
-          from: admin
-      }), "The library has already been initialized.");
+        // Test original logic contract
+        await expectRevert(web3tx(rTokenLogic.initialize, "rTokenLogic.initialize second time")(compoundAS.address, {
+            from: admin
+        }), "The library has already been initialized.");
 
-      await expectRevert(web3tx(rTokenLogic.updateCode, "rTokenLogic.updateCode from non-owner")(compoundAS.address, {
-          from: customer1
-      }), "Ownable: caller is not the owner");
+        await expectRevert(web3tx(rTokenLogic.updateCode, "rTokenLogic.updateCode from non-owner")(compoundAS.address, {
+            from: customer1
+        }), "Ownable: caller is not the owner");
 
-      // await expectRevert(web3tx(rTokenLogic.updateCode, "rTokenLogic.updateCode from owner")(compoundAS.address, {
-      //   from: admin
-      // }), "The library is locked. No direct 'call' is allowed.");
+        // await expectRevert(web3tx(rTokenLogic.updateCode, "rTokenLogic.updateCode from owner")(compoundAS.address, {
+        //   from: admin
+        // }), "The library is locked. No direct 'call' is allowed.");
 
-      // Test original proxy contract
-      await expectRevert(web3tx(rToken.initialize, "rToken.initialize (original)")(compoundAS.address, {
-          from: admin
-      }), "The library has already been initialized.");
+        // Test original proxy contract
+        await expectRevert(web3tx(rToken.initialize, "rToken.initialize (original)")(compoundAS.address, {
+            from: admin
+        }), "The library has already been initialized.");
 
-      await expectRevert(web3tx(rToken.updateCode, "rToken.updateCode (original)")(compoundAS.address, {
-          from: customer1
-      }), "Ownable: caller is not the owner");
+        await expectRevert(web3tx(rToken.updateCode, "rToken.updateCode (original)")(compoundAS.address, {
+            from: customer1
+        }), "Ownable: caller is not the owner");
 
-      // Deploy new rToken logic/library contract
-      const newRTokenLogic = await web3tx(RToken.new, "RToken.new")(
-          {
+        // Deploy new rToken logic/library contract
+        const newRTokenLogic = await web3tx(RToken.new, "RToken.new")(
+            {
+                from: admin
+            });
+        // Perform the upgrade
+        await web3tx(rToken.updateCode, "rToken.updateCode")(newRTokenLogic.address, {
             from: admin
         });
-      // Perform the upgrade
-      await web3tx(rToken.updateCode, "rToken.updateCode")(newRTokenLogic.address, {
-        from: admin
-      });
 
-      // Call initialize first time
-      await web3tx(newRTokenLogic.initialize, "rTokenLogic.initialize first time")(compoundAS.address, {
-          from: admin
-      })
+        // Call initialize first time
+        await web3tx(newRTokenLogic.initialize, "rTokenLogic.initialize first time")(compoundAS.address, {
+            from: admin
+        });
 
-      // Test new logic contract
-      await expectRevert(web3tx(newRTokenLogic.initialize, "rTokenLogic.initialize second time")(compoundAS.address, {
-          from: admin
-      }), "The library has already been initialized.");
+        // Test new logic contract
+        await expectRevert(web3tx(newRTokenLogic.initialize, "rTokenLogic.initialize second time")(compoundAS.address, {
+            from: admin
+        }), "The library has already been initialized.");
 
-      await expectRevert(web3tx(newRTokenLogic.updateCode, "rTokenLogic.updateCode from non-owner")(compoundAS.address, {
-          from: customer1
-      }), "Ownable: caller is not the owner");
+        await expectRevert(web3tx(newRTokenLogic.updateCode, "rTokenLogic.updateCode from non-owner")(compoundAS.address, {
+            from: customer1
+        }), "Ownable: caller is not the owner");
 
-      // await expectRevert(web3tx(rTokenLogic.updateCode, "rTokenLogic.updateCode from owner")(compoundAS.address, {
-      //   from: admin
-      // }), "The library is locked. No direct 'call' is allowed.");
+        // await expectRevert(web3tx(rTokenLogic.updateCode, "rTokenLogic.updateCode from owner")(compoundAS.address, {
+        //   from: admin
+        // }), "The library is locked. No direct 'call' is allowed.");
 
-      // Test new proxy contract
-      await expectRevert(web3tx(rToken.initialize, "rToken.initialize (original)")(compoundAS.address, {
-          from: admin
-      }), "The library has already been initialized.");
+        // Test new proxy contract
+        await expectRevert(web3tx(rToken.initialize, "rToken.initialize (original)")(compoundAS.address, {
+            from: admin
+        }), "The library has already been initialized.");
 
-      await expectRevert(web3tx(rToken.updateCode, "rToken.updateCode (original)")(compoundAS.address, {
-          from: customer1
-      }), "Ownable: caller is not the owner");
+        await expectRevert(web3tx(rToken.updateCode, "rToken.updateCode (original)")(compoundAS.address, {
+            from: customer1
+        }), "Ownable: caller is not the owner");
 
-    })
+    });
 
     it("#17 storage continuity during upgrade", async () => {
         // Test to ensure storage is not disrupted by performing an upgrade.
``

hellwolf commented:
cl0sing  #3
